### PR TITLE
LPS-185222 Add new files to the "new" store area

### PR DESCRIPTION
**Pull request by adolfopa**
# What is this about?

https://issues.liferay.com/browse/LPS-185222

Instead of adding new files to the 'live' area, add them to a new 'new' area.

The use case that makes this necessary is the following:
1. Add file A.
2. Take a DB snapshot.
3. Add file B.
4. Restore DB snapshot.

At this point, file B is an orphan, there's no reference in the database pointing to it and so it cannot be easily removed.  To avoid leaking storage space each time a backup is restored, we need a way of detecting these files that doesn't involve scanning the whole store.

By adding new files to the 'new' area, we can implement a background process (which I'll send in a followup pr) that will scan the 'new' area looking for unreferenced content. As the 'new' area will be much smaller than 'live', this process will consume less time and resources than if it had to scan the whole store.

Note that files in 'new' will be promoted to 'live' after 31 days (by default), as we won't support "going back in time" for periods longer that the configured value. That job will be performed by the same background process that cleans up 'new'. It will work similarly to the one we already have for the 'deleted' area.

### CODE REVIEW

**Comment by liferay-lima** [comment]
ci:test:sf

**Comment by liferay-lima** [comment]
ci:test:relevant

**Review by marcogalluzzi — CHANGES_REQUESTED** [review]
Very nice work! Just minor comments below...

File: portal-kernel/src/com/liferay/document/library/kernel/store/StoreArea.java [hunk 1]
```diff
@@ -73,26 +106,55 @@ public static <T, E extends Exception> T tryGetWithStoreAreas(
...
+
+				if (exception1 == null) {
+					exception1 = exception2;
+				}
 			}
 		}
 
+		if (failureCount == storeAreas.length) {
+			return ReflectionUtil.throwException(exception1);
+		}
```
[the remark this answers is not in the document]
Comment (adolfopa) [reply portal-kernel/src/com/liferay/document/library/kernel/store/StoreArea.java]: You're right. As soon as it succeeds it returns the result, so keeping track of the failures is pointless.

File: portal-kernel/src/com/liferay/document/library/kernel/store/StoreArea.java [hunk 2]
```diff
@@ -46,25 +47,57 @@ public static <E extends Exception> String[] mergeWithStoreAreas(
 			UnsafeSupplier<String[], E> supplier, StoreArea... storeAreas)
 		throws E {
 
+		int failureCount = 0;
+		Exception exception1 = null;
```
[the remark this answers is not in the document]
Comment (adolfopa) [reply portal-kernel/src/com/liferay/document/library/kernel/store/StoreArea.java:51]: Yes. Initially it was names `firstException` by SF complained, and I forgot to resort.

File: portal-kernel/src/com/liferay/document/library/kernel/store/StoreAreaAwareStoreWrapper.java [hunk 3]
```diff
@@ -40,39 +43,62 @@ public void addFile(
...
+				});
+		}
+		else {
+			Store store = _storeSupplier.get();
 
-		store.addFile(
-			companyId, repositoryId, fileName, versionLabel, inputStream);
+			store.addFile(
+				companyId, repositoryId, fileName, versionLabel, inputStream);
+		}
```
[the remark this answers is not in the document]
Comment (adolfopa) [reply portal-kernel/src/com/liferay/document/library/kernel/store/StoreAreaAwareStoreWrapper.java]: Done. Lifted all variables up consistently in all methods.

**Comment by adolfopa** [comment]
@marcogalluzzi I squashed all requested changes in https://github.com/liferay-lima/liferay-portal/pull/4356/commits/6f9c4b66ec6bb8143c818116fc3d2b5cf2ff86c8 and https://github.com/liferay-lima/liferay-portal/pull/4356/commits/5bc55b16d45db378b5f9ec71e486c5e5f2ef874c.

**Comment by adolfopa** [comment]
ci:test:sf

**Review by marcogalluzzi — APPROVED** [review]

**Comment by Hanlf** [comment]
ci:test:relevant

**Comment by adolfopa** [comment]
See https://github.com/liferay-lima/liferay-portal/pull/4361

**Pull request closed without merging**